File: genonaut/api/services/lora_model_service.py
"""LoRA model service for business logic operations."""

from typing import List, Optional, Tuple, Dict, Any
from uuid import UUID
from sqlalchemy.orm import Session

from genonaut.db.schema import LoraModel, CheckpointModel
from genonaut.api.repositories.lora_model_repository import LoraModelRepository
from genonaut.api.repositories.checkpoint_model_repository import CheckpointModelRepository
from genonaut.api.exceptions import EntityNotFoundError


class LoraModelService:
    """Service class for LoRA model business logic."""

    def __init__(self, db: Session):
        """Initialize service with database session.

        Args:
            db: SQLAlchemy database session
        """
        self.repository = LoraModelRepository(db)
        self.checkpoint_repository = CheckpointModelRepository(db)
        self.db = db

    def get_all(self) -> List[LoraModel]:
        """Get all LoRA models sorted by rating descending.

        Returns:
            List of all LoRA models sorted by rating (highest first)
        """
        return self.repository.get_all()

    def get_by_id(self, id: UUID) -> LoraModel:
        """Get LoRA model by ID.

        Args:
            id: LoRA model UUID

        Returns:
            LoraModel instance

        Raises:
            EntityNotFoundError: If LoRA model not found
        """
        model = self.repository.get_by_id(id)
        if model is None:
            raise EntityNotFoundError("LoraModel", id)
        return model

    def get_by_compatible_architecture(self, architecture: str) -> List[LoraModel]:
        """Get LoRA models by compatible architecture.

        Args:
            architecture: Compatible architecture (e.g., 'sd1', 'sdxl')

        Returns:
            List of LoRA models compatible with the architecture sorted by rating
        """
        return self.repository.get_by_compatible_architecture(architecture)

    def get_paginated_with_compatibility(
        self,
        page: int = 1,
        page_size: int = 10,
        checkpoint_id: Optional[str] = None
    ) -> Tuple[List[Dict[str, Any]], int, int]:
        """Get paginated LoRA models with compatibility and optimality flags.

        Args:
            page: Page number (1-indexed)
            page_size: Number of items per page
            checkpoint_id: UUID of checkpoint to check compatibility against

        Returns:
            Tuple of (list of enriched LoRA model dicts, total count, total pages)
        """
        checkpoint = None
        if checkpoint_id:
            checkpoint = self.checkpoint_repository.get_by_id(UUID(checkpoint_id))

        # Get paginated models
        models, total = self.repository.get_paginated(
            page=page,
            page_size=page_size
        )

        # Enrich with compatibility/optimality flags
        enriched_models = []
        for model in models:
            enriched = {
                'model': model,
                'is_compatible': self._check_compatibility(model, checkpoint) if checkpoint else None,
                'is_optimal': self._check_optimality(model, checkpoint) if checkpoint else None,
            }
            enriched_models.append(enriched)

        total_pages = (total + page_size - 1) // page_size

        return enriched_models, total, total_pages

    def _check_compatibility(self, lora: LoraModel, checkpoint: Optional[CheckpointModel]) -> bool:
        """Check if LoRA is compatible with checkpoint.

        Args:
            lora: LoRA model to check
            checkpoint: Checkpoint model to check against

        Returns:
            True if compatible, False otherwise
        """
        if not checkpoint or not checkpoint.architecture:
            return False

        if not lora.compatible_architectures:
            return False

        # Case-insensitive partial match
        checkpoint_arch = checkpoint.architecture.lower()
        lora_archs = lora.compatible_architectures.lower()

        return checkpoint_arch in lora_archs

    def _check_optimality(self, lora: LoraModel, checkpoint: Optional[CheckpointModel]) -> bool:
        """Check if LoRA is optimal for checkpoint.

        Args:
            lora: LoRA model to check
            checkpoint: Checkpoint model to check against

        Returns:
            True if optimal, False otherwise
        """
        if not checkpoint or not checkpoint.family:
            return False

        if not lora.optimal_checkpoints or len(lora.optimal_checkpoints) == 0:
            return False

        # Case-insensitive check if family is in optimal_checkpoints list
        checkpoint_family = checkpoint.family.lower()
        optimal_checkpoints = [cp.lower() for cp in lora.optimal_checkpoints]

        return checkpoint_family in optimal_checkpoints

File: frontend/src/services/lora-model-service.ts
import { ApiClient } from './api-client'
import type { ApiLoraModel, ApiLoraModelListResponse } from '../types/api'
import type { LoraModel, PaginatedLoraModels } from '../types/domain'

export interface LoraModelQueryParams {
  page?: number
  pageSize?: number
  checkpointId?: string
}

export class LoraModelService {
  private readonly api: ApiClient

  constructor(api: ApiClient) {
    this.api = api
  }

  async getAll(): Promise<LoraModel[]> {
    const response = await this.api.get<ApiLoraModelListResponse>('/api/v1/lora-models/')
    return response.items.map(this.transformModel)
  }

  async getPaginated(params: LoraModelQueryParams = {}): Promise<PaginatedLoraModels> {
    const queryParams = new URLSearchParams()

    if (params.page) queryParams.append('page', params.page.toString())
    if (params.pageSize) queryParams.append('page_size', params.pageSize.toString())
    if (params.checkpointId) queryParams.append('checkpoint_id', params.checkpointId)

    const response = await this.api.get<ApiLoraModelListResponse>(
      `/api/v1/lora-models/?${queryParams.toString()}`
    )

    return {
      items: response.items.map(this.transformModel),
      total: response.total,
      pagination: response.pagination ? {
        page: response.pagination.page,
        pageSize: response.pagination.page_size,
        total: response.pagination.total,
        totalPages: response.pagination.total_pages,
      } : {
        page: 1,
        pageSize: 10,
        total: response.total,
        totalPages: Math.ceil(response.total / 10),
      }
    }
  }

  async getById(id: string): Promise<LoraModel> {
    const apiModel = await this.api.get<ApiLoraModel>(`/api/v1/lora-models/${id}`)
    return this.transformModel(apiModel)
  }

  private transformModel(apiModel: ApiLoraModel): LoraModel {
    return {
      id: apiModel.id,
      path: apiModel.path,
      filename: apiModel.filename,
      name: apiModel.name,
      version: apiModel.version,
      compatibleArchitectures: apiModel.compatible_architectures,
      family: apiModel.family,
      description: apiModel.description,
      rating: apiModel.rating,
      tags: apiModel.tags,
      triggerWords: apiModel.trigger_words,
      optimalCheckpoints: apiModel.optimal_checkpoints,
      modelMetadata: apiModel.model_metadata,
      createdAt: apiModel.created_at,
      updatedAt: apiModel.updated_at,
      isCompatible: apiModel.is_compatible,
      isOptimal: apiModel.is_optimal,
    }
  }
}
